server/website/website/views.py

```diff
@@ -486,6 +488,54 @@ def handle_result_files(session, files, execution_times=None):
     # Load the contents of the controller's summary file
     summary = JSONUtil.loads(files['summary'])
 
+    dbms_id = session.dbms.pk
+    udm_before = {}
+    udm_after = {}
+    if ('user_defined_metrics' not in files) and (USER_DEFINED_TARGET == session.target_objective):
+        return HttpResponse('ERROR: user defined target objective is not uploaded!')
+    # User defined metrics
+    udm = {}
+    if 'user_defined_metrics' in files:
+        udm = JSONUtil.loads(files['user_defined_metrics'])
+    if len(udm) > 0:
+        udm_target = udm['target_objective']
+        udm_not_target = udm['metrics']
+        udm_all = copy.deepcopy(udm_not_target)
+        if (udm_target is None) and (USER_DEFINED_TARGET == session.target_objective):
+            return HttpResponse('ERROR: user defined target objective is not uploaded!')
+        if udm_target is not None:
+            udm_all.update(udm_target)
+        if not target_objectives.udm_registered(dbms_id):
+            target_objectives.register_udm(dbms_id, udm_all)
+        for name, info in udm_all.items():
+            udm_name = 'udm.' + name
+            udm_before[name] = 0
+            udm_after[name] = info['value']
+            if MetricCatalog.objects.filter(dbms=session.dbms, name=udm_name).exists():
+                continue
+            udm_catalog = MetricCatalog.objects.create(dbms=session.dbms,
+                                                       name=udm_name,
+                                                       vartype=info['type'],
+                                                       scope='global',
+                                                       metric_type=MetricType.STATISTICS)
+            udm_catalog.summary = 'user defined metric, not target objective'
+            udm_catalog.save()
+        if udm_target is not None:
+            target_name = 'udm.' + list(udm_target.keys())[0]
+            pprint_name = 'udf.' + list(udm_target.keys())[0]
```

Comment:
why named 'udf'?

Reply:
udf is "user defined function", which means a user defined metric as a target objective. 

It will show both udf and udm. For example, udf.latency and udm.latency. By default, they should be same. However, you can modify the TargetObjectiveClass to calculate udf from udm.  e.g., udm is 1000 ms,  you can set udf = udm / 1000 = 1, and change the unit to second.

Reply:
OK. But line 524 and line 525 are inconsistent. Is that meant to be so?

Reply:
Yes. udf is derived from a udm.